### PR TITLE
[CARBONDATA-2209] Fixed rename table with partitions not working issue and batch_sort and no_sort with partition table issue

### PR DESCRIPTION
**Pull request by ravipesala**
Problems
After table rename on partitions table, it returns empty data upon querying.
Batch sort and no sort loading is not working on partition table.

Solution
During alter rename table we should also set the new table location as we are changing the folder name.
Batch sort and no sort should go through the new partition load flow
And also restricted set location through partition spec as it is high risk as of now.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 
 - [x] Any backward compatibility impacted?
 
 - [x] Document update required?

 - [x] Testing done
        Tests added
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2699/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3943/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2700/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3945/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3706/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3982/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2738/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3739/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3999/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2755/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2765/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4009/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4018/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2774/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/SegmentIndexFileStore.java [hunk 1]
```diff
@@ -108,11 +109,20 @@ public void readAllIIndexOfSegment(SegmentFileStore segmentFileStore, SegmentSta
...
         }
+        String mergeFileName = locations.getValue().getMergeFileName();
         for (String indexFile : locations.getValue().getFiles()) {
           CarbonFile carbonFile = FileFactory
               .getCarbonFile(location + CarbonCommonConstants.FILE_SEPARATOR + indexFile);
-          if (carbonFile.exists()) {
+          if (carbonFile.exists() && !indexFiles.contains(carbonFile.getAbsolutePath())) {
             carbonIndexFiles.add(carbonFile);
+            indexFiles.add(carbonFile.getAbsolutePath());
+          } else if (mergeFileName != null) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/SegmentIndexFileStore.java]: Can move mergeFilename logic out of loop

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 2]
```diff
@@ -38,85 +46,158 @@
...
+      // info in the index file and therefore blocklet info need to be read from the file footer
+      // in the carbondata file
+      fileStore.readAllIndexAndFillBolckletInfo(segmentPath);
+    } else {
+      fileStore.readAllIIndexOfSegment(segmentPath);
+    }
+    Map<String, byte[]> indexMap = fileStore.getCarbonIndexMap();
+    writeMergeIndexFile(indexFileNamesTobeAdded, segmentPath, indexMap);
+    for (CarbonFile indexFile : indexFiles) {
+      indexFile.delete();
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java]: deletion should be postponed as parallel read can happen

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableRenameCommand.scala [hunk 3]
```diff
@@ -138,6 +147,27 @@ private[sql] case class CarbonAlterTableRenameCommand(
           sys.error(s"Folder rename failed for table $oldDatabaseName.$oldTableName")
         }
       }
+      val updatedParts = updatePartitionLocations(
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableRenameCommand.scala]: need to check once how hive managed table rename works? whether it renames location or not? If not how table creation with old table name works?

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/SegmentIndexFileStore.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/SegmentIndexFileStore.java]: ok

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java (hunk 2 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableRenameCommand.scala (hunk 3 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableRenameCommand.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4047/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2801/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4048/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2802/

**Comment by gvramana** [comment]
LGTM

**Pull request closed without merging**